### ISSUE TITLE
Strange behavior --- BadPointerLoad

### ISSUE THREAD

**Issue by weaversa** [GaloisInc/saw-script#257]
Here is my attempt at a minimal example for a potential `saw` bug. Any help would be greatly appreciated.

test.c
```C
#define SIZE 2

typedef struct {
  int x[SIZE];
  int y[SIZE];
} two_t;

void test(int x[SIZE]) {
}

void broken(two_t *a, two_t *b) {
  if(a->x[0] == 0)
    test(b->x);
}
```

test.saw
```Haskell
let alloc_init ty v = do {
  p <- crucible_alloc ty;
  crucible_points_to p v;
  return p;
};

let ptr_to_fresh n ty = do {
  x <- crucible_fresh_var n ty;
  p <- alloc_init ty (crucible_term x);
  return (x, p);
};

let test_setup = do {
  (x, p_x) <- ptr_to_fresh "x" (llvm_array 2 (llvm_int 32));

  crucible_execute_func [p_x];

  crucible_points_to p_x (crucible_term {{ x }});
};

let broken_setup = do {
  (a, p_a) <- ptr_to_fresh "a" (llvm_struct "struct.two_t");
  (b, p_b) <- ptr_to_fresh "b" (llvm_struct "struct.two_t");

  crucible_execute_func [p_a, p_b];

  crucible_points_to p_b (crucible_term {{ b }});
};

test_m <- llvm_load_module "test.bc";

test_ov   <- crucible_llvm_verify test_m "test" [] false test_setup z3;
broken_ov <- crucible_llvm_verify test_m "broken" [test_ov] false broken_setup z3;
```

result:
```
$ clang --version
Apple LLVM version 9.0.0 (clang-900.0.39.2)
Target: x86_64-apple-darwin17.3.0
$ saw --version
0.2 (c54609f)
$ clang -emit-llvm -Wall test.c -c -o test.bc && saw test.saw
Loading file "test.saw"
Proof succeeded! @test
Registering override for `test`
user error ("crucible_llvm_verify" (test.saw:33:14):
BadPointerLoad "Invalid memory load: address (4, 0x0:[64]) at type {[2 x i32], [2 x i32]}")
```

If I change this
`(x, p_x) <- ptr_to_fresh "x" (llvm_array 2 (llvm_int 32));`
to this
`(x, p_x) <- ptr_to_fresh "x" (llvm_array 3 (llvm_int 32));`
I get:
```
$ clang -emit-llvm -Wall test.c -c -o test.bc && saw test.saw
Loading file "test.saw"
Proof succeeded! @test
Registering override for `test`
Subgoal failed: @broken safety assertion: Invalid memory load: address (4, 0x0:[64]) at type {[2 x i32], [2 x i32]}
SolverStats {solverStatsSolvers = fromList ["SBV->Z3"], solverStatsGoalSize = 62}
----------Counterexample----------
("a",([0,0],[0,0]))
("b",([0,0],[0,0]))
user error ("crucible_llvm_verify" (test.saw:33:14):
Proof failed.)
```

Or if I change this
`(x, p_x) <- ptr_to_fresh "x" (llvm_array 2 (llvm_int 32));`
to this
`(x, p_x) <- ptr_to_fresh "x" (llvm_array 4 (llvm_int 32));`
I get:
```
$ clang -emit-llvm -Wall test.c -c -o test.bc && saw test.saw
Loading file "test.saw"
Proof succeeded! @test
Registering override for `test`
Proof succeeded! @broken
```

Or if I change this
`(x, p_x) <- ptr_to_fresh "x" (llvm_array 2 (llvm_int 32));`
to this
`(x, p_x) <- ptr_to_fresh "x" (llvm_array 5 (llvm_int 32));`
I get:
```
$ clang -emit-llvm -Wall test.c -c -o test.bc && saw test.saw
Loading file "test.saw"
Proof succeeded! @test
Registering override for `test`
All overrides failed: [BadPointerLoad "Invalid memory load: address (4, 0x0:[64]) at type [5 x i32]"]
in broken at internal
  When calling test
  In broken at internal
Symbolic simulation failed along some paths!
Subgoal failed: @broken safety assertion: failed path error
SolverStats {solverStatsSolvers = fromList ["SBV->Z3"], solverStatsGoalSize = 37}
----------Counterexample----------
("a",([0,0],[0,0]))
user error ("crucible_llvm_verify" (test.saw:33:14):
Proof failed.)
```

Or, if I change this
```
typedef struct {
  int x[SIZE];
  int y[SIZE];
} two_t;
```
to this
```
typedef struct {
  int x[SIZE];
  //int y[SIZE];
} two_t;
```
I get:
```
$ clang -emit-llvm -Wall test.c -c -o test.bc && saw test.saw
Loading file "test.saw"
Proof succeeded! @test
Registering override for `test`
Proof succeeded! @broken
```

**Comment by brianhuffman** [comment]
I added some temporary debugging output to SAW, to get a memory dump when the BadPointerLoad exception is thrown:
```
Base memory
  Allocations:
    merge
      Condition:
        bvEq 0x0:[1] (bvIte (bvEq 0x0:[32] c@11:bv) 0x1:[1] 0x0:[1])
      True Branch:

      False Branch:

    HeapAlloc 5 0x10:[64] <malloc>
    HeapAlloc 4 0x10:[64] <malloc>
    GlobalAlloc 3 0x0:[64] test
    GlobalAlloc 2 0x0:[64] llvm.dbg.declare
    GlobalAlloc 1 0x0:[64] broken
  Writes:
    merge
      Condition:
        bvEq 0x0:[1] (bvIte (bvEq 0x0:[32] c@11:bv) 0x1:[1] 0x0:[1])
      True Branch:

      False Branch:
        *(5, 0x0:[64]) := [c@19:bv
                          ,c@20:bv]
    *(7, 0x0:[64]) := (5, 0x0:[64])
    *(6, 0x0:[64]) := (4, 0x0:[64])
    *(4, 0x0:[64]) := {base+0 = [c@11:bv
                                ,c@12:bv]
                      ,base+8 = [c@13:bv,c@14:bv]}
    *(5, 0x0:[64]) := {base+0 = [c@7:bv
                                ,c@8:bv]
                      ,base+8 = [c@9:bv,c@10:bv]}
user error ("crucible_llvm_verify" (/Users/huffman/Documents/saw/issue257.saw:39:14-39:34):
BadPointerLoad "Invalid memory load: address (5, 0x0:[64]) at type {[2 x i32], [2 x i32]}")
```

**Comment by brianhuffman** [comment]
I was able to shrink the example a bit further. The arrays are not essential. What is important is that there is a write to *part* of a struct that is guarded by a conditional. (The origin of the write is the `crucible_points_to` statement in the override for `test`.)
```
typedef struct {
	int x;
	int y;
} two_t;

void test(int *x) {
}

void broken(two_t *a, two_t *b) {
	if(a->x == 0)
		test(&(b->x));
}
```

**Comment by brianhuffman** [comment]
I know exactly what the problem is now. It turns out I introduced a bug in revision GaloisInc/crucible@e242ec3b when I was fixing a performance problem (#198). I'm working on designing a fix now.

**Comment by brianhuffman** [comment]
Fixed in GaloisInc/crucible@ec511cc0 and GaloisInc/saw-script@ae39976e.

**Comment by weaversa** [comment]
Thank you thank you thank you!!!
